examples/siamese_network/siamese_network.py

```diff
@@ -0,0 +1,284 @@
+        self.grouped_examples = {}
+        for i in range(0, 10):
+            self.grouped_examples[i] = np.where((np_arr == i))[0]
+
+    def __len__(self):
+        return self.data.shape[0]
+
+    def __getitem__(self, index):
+        """
+        For every sample in the batch we select 3 images. First one is the anchor image
+        which is the image obtained from the current index. We also obtain the label of
+        anchor image.
+
+        Now we select two random images, one belonging to the same class as that of the
+        anchor image (named as positive_image) and the other belonging to a different class
+        than that of the anchor image (named as negative_image). We return the anchor image,
+        positive image, negative image and anchor label.
+        """
+
+        # obtain the anchor image
+        anchor_image = self.data[index].clone().float()
+
+        # obtain the class label of the anchor image
+        anchor_label = self.dataset.targets[index].clone().float()
+
+        # pick some random class for the first image
+        selected_class = random.randint(0, 9)
+
+        # pick a random index for the first image in the grouped indices based of the label
+        # of the class
+        random_index_1 = random.randint(0, self.grouped_examples[selected_class].shape[0] - 1)
+
+        while random_index_1 != anchor_label:
+            random_index_1 = random.randint(0, self.grouped_examples[selected_class].shape[0] - 1)
+
+        # pick the index to get the first image
+        index_1 = self.grouped_examples[selected_class][random_index_1]
+
+        # get the first image
+        positive_image = self.data[index_1].clone().float()
+
+        random_index_2 = random.randint(0, self.grouped_examples[selected_class].shape[0] - 1)
+
+        # ensure that the index of the second image isn't the same as the first image
+        while random_index_2 == anchor_label:
+            random_index_2 = random.randint(0, self.grouped_examples[selected_class].shape[0] - 1)
+
+        # pick the index to get the second image
+        index_2 = self.grouped_examples[selected_class][random_index_2]
+
+        # get the second image
+        negative_image = self.data[index_2].clone().float()
+
+        return anchor_image, positive_image, negative_image, anchor_label
+
+
+def train(model, device, optimizer, train_loader, lr_scheduler, log_interval, max_epochs):
+
+    criterion = nn.TripletMarginLoss()
+
+    # define model training step
+    def train_step(engine, batch):
+        model.train()
+        anchor_image, positive_image, negative_image, anchor_label = batch
+        anchor_image = anchor_image.to(device)
+        positive_image, negative_image = positive_image.to(device), negative_image.to(device)
+        anchor_label = anchor_label.to(device)
+        optimizer.zero_grad()
+        anchor_out = model(anchor_image)
+        positive_out = model(positive_image)
+        negative_out = model(negative_image)
```

Comment:
Why we are doing 3 forward passes here ? Can't we group them into a single one ?